### PR TITLE
[Data] Report first block size per `Operator`

### LINKED ISSUE #39647: [data] Better reporting for block sizes
### Description

Unexpected large in-memory block sizes are a common source of instability or other performance issues. To help with debugging these issues, Ray Data should:

1) log as INFO actual in-memory block size once per operator, this can be done in the executor when the first block for the operator is produced.
2) log as WARNING if the in-memory block size is significantly larger than the target max block size (e.g., >2x larger), and recommend manually increasing the number of blocks via the parallelism setting to avoid issues

### Use case

_No response_

### PR DESCRIPTION
**Pull request by scottjlee**
## Why are these changes needed?
For each `Operator`, logs an info message with the in-memory size of the first block that is produced. If the size of this block greatly exceeds the target max block size (currently configured at 2x), logs a warning message.

## Related issue number

Closes #39647

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/tests/block_batching/test_util.py [hunk 1]
```diff
@@ -9,15 +9,14 @@
...
-    Queue,
     _calculate_ref_hits,
     blocks_to_batches,
     collate,
     finalize_batches,
     format_batches,
     resolve_block_refs,
 )
-from ray.data._internal.util import make_async_gen
+from ray.data._internal.util import Queue, make_async_gen
```
Comment (scottjlee) [inline python/ray/data/tests/block_batching/test_util.py:19]: fixes breaking test import

File: python/ray/data/_internal/execution/streaming_executor_state.py [hunk 2]
```diff
@@ -259,6 +266,31 @@ def _queue_memory_usage(self, queue: Deque[RefBundle]) -> int:
...
+        target_max_block_size = DataContext.get_current().target_max_block_size
+
+        if self.first_block_size_bytes > (
+            BLOCK_SIZE_TO_MAX_TARGET_RATIO * target_max_block_size
+        ):
+            logger.get_logger().warning(
+                f"{self.op.name} in-memory block size of "
+                f"{(self.first_block_size_bytes / 2**20):.2f} MB is significantly "
+                f"larger than the maximium target block size of "
+                f"{(target_max_block_size / 2**20):.2f} MB."
```
Comment (amogkam) [inline python/ray/data/_internal/execution/streaming_executor_state.py:286]: if a user sees this warning message, what are they expected to do?

**Review by amogkam — COMMENTED** [review]
+1 to @raulchen.

This information is useful for when the Ray team is debugging, but if a user sees this message I'm not sure what they are expected to do.

I think this information is more useful to save as part of the persisted dataset stats (and we can even go one step further and track min output block size, average output block size, max output block size). This way users can easily send logs for debugging.

But I don't think this is helpful to log on stdout.

**Comment by scottjlee** [comment]
@amogkam @raulchen - discussed with @ericl, ideally we would want the user to look at the warning and realize they should try adjusting parallelism in order to reduce block size. we can add a paragraph in the [Performance Tips](https://docs.ray.io/en/latest/data/performance-tips.html#tuning-read-parallelism) page describing this -- i think we discuss it in the ["Tuning read parallelism"](https://docs.ray.io/en/latest/data/performance-tips.html#tuning-read-parallelism) section, but there's no clear course of action.

File: python/ray/data/_internal/execution/streaming_executor_state.py [hunk 3]
```diff
@@ -259,6 +266,31 @@ def _queue_memory_usage(self, queue: Deque[RefBundle]) -> int:
...
+            )
+        else:
+            logger.get_logger().info(
+                f"{self.op.name} in-memory block size: "
+                f"{(self.first_block_size_bytes / 2**20):.2f} MB"
+            )
```
[the remark this answers is not in the document]
Comment (scottjlee) [reply python/ray/data/_internal/execution/streaming_executor_state.py:295]: @ericl what was the original scenario where this issue appeared? The block splitting should be ensuring we don't get block sizes larger than ``target_max_block_size`, are there operators where this is not being handled?

**Comment by ericl** [comment]
I'm not sure. One reason to add these logs is to surface cases where block
splitting didn't work for whatever reason.

**Comment by scottjlee** [comment]
It sounds like from this discussion, we want to add this as an "insurance policy," i.e. users really shouldn't be running into this issue, but in case they do, we log a warning with a suggestion to increase parallelism. Since the current changes will only emit a warning if the block size exceeds the configured target size (very rare), there's little downside to including this right (i.e. no excessive spam, it's printed only when truly needed).

Since we only log the warning when the block size exceeds the configured target size (which should rarely be happening), and we always log the info to the data-specific log file, I think users wouldn't be getting spammed. Later, we can add additional statistics like min/avg/max block size to the dashboard, like Amog suggested.

**Comment by scottjlee** [comment]
It sounds like from this discussion, we want to add this as an "insurance policy," i.e. users really shouldn't be running into this issue, but in case they do, we log a warning with a suggestion to increase parallelism. Since the current changes will only emit a warning if the block size exceeds the configured target size (very rare), there's little downside to including this (i.e. no excessive spam, it's printed only when truly needed), so I think this is a beneficial addition? @amogkam @raulchen 

Since we only log the warning when the block size exceeds the configured target size (which should rarely be happening), and we always log the info to the data-specific log file, I think users wouldn't be getting spammed. Later, we can add additional statistics like min/avg/max block size to the dashboard, like Amog suggested.

**Comment by ericl** [comment]
Yes, and thinking about this more another scenario is if the user accidentally returns a single humongous row or something like this, which isn't an uncommon error when working with tensor data.

Even just playing around with a couple examples, I found a bug where `map_batches` doesn't seem to split blocks into the right size, and without this kind of log it would be pretty hard to identify these sort of issues.

**Comment by raulchen** [comment]
Since this is for dev to debugging perf, only logging to data logs makes more sense.

**Comment by scottjlee** [comment]
> Since this is for dev to debugging perf, only logging to data logs makes more sense.

@raulchen Do you suggest we move both the log and warn to data log only? I think showing the warning in stdout still makes sense, since this is a pretty large issue that users should be aware of without having to look at data specific logs.

**Comment by raulchen** [comment]
Keeping the warning in stdout is fine, as long as we can the message clear. I don't think we should suggest users to increase the parallelism. Because this issue can only happen when (1) one single row is bigger than the target block size; 2) there is a bug in Ray Data.

**Comment by ericl** [comment]
I think this has to be a high severity level, if we think it indicates a big. We might even raise an exception in the future if this happens.

However, I don't think today we can raise an exception, since UDFs can return large blocks in map batches and these aren't split. The best we can do is warn about this.

**Comment by scottjlee** [comment]
Got it, @raulchen @ericl updated the PR to match discussion from above, ready for another look.

File: python/ray/data/tests/test_streaming_executor.py [hunk 4]
```diff
@@ -684,6 +687,56 @@ def test_execution_allowed_nothrottle():
     )
 
 
+def test__check_first_block_size():
```
Comment (raulchen) [inline python/ray/data/tests/test_streaming_executor.py:690]: is it a typo to have 2 underscores?

**Review by raulchen — APPROVED** [review]

**Comment by scottjlee** [comment]
Will wait for https://github.com/ray-project/ray/pull/40173 to be merged, then we can also emit better warnings based on the block size metrics available per Operator.

**Comment by scottjlee** [comment]
Original issue closed, no longer needed.

**Pull request closed without merging**